ldm/models/diffusion/shared_invokeai_diffusion.py

```diff
@@ -294,25 +298,29 @@ def _threshold(self, threshold, warmup, latents: torch.Tensor, sigma) -> torch.T
         if self.debug_thresholding:
             std, mean = [i.item() for i in torch.std_mean(latents)]
             outside = torch.count_nonzero((latents < -current_threshold) | (latents > current_threshold))
-            print(f"\nThreshold: 𝜎={sigma.item()} threshold={current_threshold:.3f} (of {threshold:.3f})\n"
+            print(f"\nThreshold: %={percent_through} threshold={current_threshold:.3f} (of {threshold:.3f})\n"
                   f"  | min, mean, max = {minval:.3f}, {mean:.3f}, {maxval:.3f}\tstd={std}\n"
                   f"  | {outside / latents.numel() * 100:.2f}% values outside threshold")
 
         if maxval < current_threshold and minval > -current_threshold:
             return latents
 
         if maxval > current_threshold:
+            latents = torch.clone(latents)
             maxval = np.clip(maxval * scale, 1, current_threshold)
+            latents[latents > maxval] = torch.rand_like(latents[latents > maxval]) * maxval
```

Comment:
can't use rand_like as-is on mps. the generator needs to run on CPU and the resulting tensor moved.

Reply:
as for the noising here... it's going to replace these thresholded areas with noise scaled to `maxval`, still zero-centered?

I'd have expected you want it added to maxval or maxval × scale or current_threshold or some such.

Reply:
> can't use rand_like as-is on mps. the generator needs to run on CPU and the resulting tensor moved.

Can you clue me in on how to do this?


> as for the noising here... it's going to replace these thresholded areas with noise scaled to `maxval`, still zero-centered?
> 
> I'd have expected you want it added to maxval or maxval × scale or current_threshold or some such.

I was using the code that @pbaylies wrote. I'd love for them to chime in here.

Reply:
i'm quite certain the latents go negative too - so admittedly not paying too close attention to exactly what the code is doing this does like it's clipping the +ve but not clipping the -ve, which seems wrong.

Reply:
It handles both cases, see https://github.com/invoke-ai/InvokeAI/pull/2479/files#diff-f4fa0f59536c4276d8f6a0c4b4dcec17e99382af015315c0144b9cc1c6a48132R316

Reply:
i see, yes. disregard my comment.

Reply:
Just pulled in code from `ldm.generate` to do the MPS version of `rand_like`. Is that even necessary or does that override the default `torch.rand_like`?

Reply:
@keturn I'd love your input here so I can understand what's going on in `generate.py`.

Reply:
I investigated and `generate.py` fixes all sorts of torch functions. By the time we get to generating random numbers for thresholding, this function is safe for MPS as-is.